e2e/txpool_test.go

```diff
@@ -0,0 +1,125 @@
+package e2e
+
+import (
+	"context"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/0xPolygon/minimal/crypto"
+	"github.com/0xPolygon/minimal/e2e/framework"
+	"github.com/0xPolygon/minimal/helper/tests"
+	txpoolOp "github.com/0xPolygon/minimal/txpool/proto"
+	"github.com/0xPolygon/minimal/types"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTxPool_TransactionCoalescing(t *testing.T) {
+	// Test scenario:
+	// Add tx with nonce 0
+	// -> Check if tx has been parsed
+	// Add tx with nonce 2
+	// -> tx shouldn't be executed, but shelved for later
+	// Add tx with nonce 1
+	// -> check if both tx with nonce 1 and tx with nonce 2 are parsed
+
+	// Predefined values
+	gasPrice := big.NewInt(10000)
+
+	referenceKey, referenceAddr := tests.GenerateKeyAndAddr(t)
+	defaultBalance := framework.EthToWei(10)
+
+	devInterval := 5
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(devInterval)
+		config.Premine(referenceAddr, defaultBalance)
+	})
+	srv := srvs[0]
+	client := srv.JSONRPC()
+
+	// Required default values
+	signer := crypto.NewEIP155Signer(100)
+
+	// TxPool client
+	clt := srv.TxnPoolOperator()
+	toAddress := types.StringToAddress("1")
+	oneEth := framework.EthToWei(1)
+
+	generateTx := func(nonce uint64) *types.Transaction {
+		signedTx, signErr := signer.SignTx(&types.Transaction{
+			Nonce:    nonce,
+			From:     referenceAddr,
+			To:       &toAddress,
+			GasPrice: gasPrice,
+			Gas:      1000000,
+			Value:    oneEth,
+			V:        1, // it is necessary to encode in rlp
+		}, referenceKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		return signedTx
+	}
+
+	generateReq := func(nonce uint64) *txpoolOp.AddTxnReq {
+		var msg *txpoolOp.AddTxnReq
+		unstakeTxn := generateTx(nonce)
+		msg = &txpoolOp.AddTxnReq{
+			Raw: &any.Any{
+				Value: unstakeTxn.MarshalRLP(),
+			},
+			From: types.ZeroAddress.String(),
+		}
+
+		return msg
+	}
+
+	// Add the transactions with the following nonce order
+	nonces := []uint64{0, 2}
+	for i := 0; i < len(nonces); i++ {
+		addReq := generateReq(nonces[i])
+
+		_, addErr := clt.AddTxn(context.Background(), addReq)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
+
+	// Mandatory sleep for the dev consensus and executor to go through the txns
+	time.Sleep(time.Duration(devInterval+1) * time.Second)
```

Comment:
I think you already implemented a better way than this on another branch, so if that's the case, just wait for that branch to be merged and use it here

Reply:
Yep. I figured out on other branches that are still awaiting PR reviews that it's better to subscribe to new blocks.

Fixed 👍 

a6edca6